### ISSUE TITLE
pixel LEDs flickering in Pixel example

### ISSUE THREAD

**Issue by imsadegh** [HomeSpan/HomeSpan#741]
I uploaded the pixel example on ESP32 WROOM and connected 16 WS2812B LEDs to pin 5.
It works well, but I notice flickering at brightness levels above 20%.
Additionally, sometimes when changing the color or brightness, some of the LEDs randomly switch to a color other than the one I set.

**Comment by HomeSpan** [comment]
Once the pixels are lit there is no communication from the ESP32.  The pixels contain all the PWM logic to stay on as long as there is power.  If you are seeing flickering this suggests a power issue.  Are you trying to power directly from the pin or are you using a separate power supply.

**Comment by imsadegh** [comment]
> Once the pixels are lit there is no communication from the ESP32. The pixels contain all the PWM logic to stay on as long as there is power. If you are seeing flickering this suggests a power issue. Are you trying to power directly from the pin or are you using a separate power supply.

I tried using both a separate power supply and the ESP pin, but got the same result.
I think if the issue were with the power, it shouldn't flicker at low brightness levels.
And checked with the FastLED library to make sure if the LEDs have a problem, and they work fine.

**Comment by HomeSpan** [comment]
Did you use the exact Pixel example, completely unmodified, except that you changed this:

```
  SPAN_ACCESSORY("Neo RGB");
    new NeoPixel_RGB(NEOPIXEL_RGB_PIN,8);
```

to this:

```
  SPAN_ACCESSORY("Neo RGB");
    new NeoPixel_RGB(5,16);
```

**Comment by imsadegh** [comment]
Yes, I only changed the pin and number of LEDs.

**Comment by HomeSpan** [comment]
Hmm.  I can't explain this.   As a check, once you set the color you should be able to disconnect pin 5 from the LED strip and nothing should change as long as the strip is still connected to power and ground.  This is an RGB device, *not* an RGBW, correct?  And to confirm, the flickering occurs only at bright levels?

**Comment by imsadegh** [comment]
Yes, and it is RGB.
No, flickering is noticeable at 20% brightness and more.

**Comment by HomeSpan** [comment]
When you compare with FastLED and set the brightness high, are the levels equivalent?  That is, with FAST LED, you are not getting flickering at, say, 50%, but doe this look as bright as 50% using the Pixel library?

Does this only occur at certain colors?  A good test is to set the color to 100% Red (fully saturated).

**Comment by imsadegh** [comment]
I haven't compared the brightness levels yet, but I'll test it. I guess that the issue occurs at 20% brightness, which isn't at the full brightness capability of the LED. Yes, it flickers in all colors.

**Comment by HomeSpan** [comment]
Can give it a try. I've not tried myself, so don't know if there are compatibility issues.

**Comment by HomeSpan** [comment]
I just re-tested my kit with the all three types of pixels and can't seem to reproduce the flickering.  I am powering the NeoPixles using an external 5V power supply (i.e. NOT connected to the ESP32, except for ground), whereas the ESP32 itself is powered by through USB connector.  I have two different 5V supplies - one is 2A and the other 1A only.  When I use the 1A the LEDs are slightly dimmer, but still no flickering (at least none that I can see).  Are you powering with an external 5V supply (with at least 1A)?

**Comment by johnwiseheart** [comment]
I'm experiencing this too with a newly bought [neopixel jewel](https://www.adafruit.com/product/2226) RGBW model. Wondering if its something to do with the RGBW protocol for the LEDs - maybe thats different to the ordinary protocol that the Pixel lib is using.

**Comment by HomeSpan** [comment]
Are you specifying RGBW when using the HomeSpan Pixel library (both RGB and RGBW are supported).

**Comment by HomeSpan** [comment]
I just looked up the jewel on Adafruit and is seems to be just RGB, not RGBW.  However, they switched to the new WS2812B LEDs, which, according to the specs I can find on the web, use slightly different [timing signals](https://cdn-shop.adafruit.com/datasheets/WS2812B.pdf) than the original WS2812.   The Pixel library allows you to specify custom timing signals.  Try adjusting the timing by calling the follows method on your Pixel object:

```
setTiming(0.40,  0.85,  0.80 , 0.45,  80);
```

**Comment by HomeSpan** [comment]
Follow-up.  I went back to the Adafruit site and realized that the jewel does come in both RGB and RGBW flavors, so make sure you set the RGBW flag to either true or false depending on your model.

**Comment by imsadegh** [comment]
> Are you powering with an external 5V supply (with at least 1A)?

Yes, it is 1A.

> setTiming(0.40,  0.85,  0.80 , 0.45,  80);

Should I set it once in setup function?

**Comment by HomeSpan** [comment]
It's a member function so you call it on each Pixel object after they are created. Once set for an instance of the object, it remains in place for all subsequent commands.

**Comment by imsadegh** [comment]
> It's a member function so you call it on each Pixel object after they are created. Once set for an instance of the object, it remains in place for all subsequent commands.

like this?
```
    V.setRange(5, 100, 1);    // sets the range of the Brightness to be from a min of 5%, to a max of 100%, in steps of 1%
    pixel = new Pixel(pin);   // creates Pixel LED on specified pin
    this->nPixels = nPixels;  // save number of Pixels in this LED Strand
    pixel->setTiming(0.40, 0.85, 0.80, 0.45, 80);
    
    update();  // manually call update() to set pixel with restored initial values
```



> Also, just to confirm, you've connected the ground from your power supply to ground on the ESP32?

Yes, the ground is common.

**Comment by HomeSpan** [comment]
Yes, that's correct.  I'm not sure if this will address the flickering issue, but it will let us see if the problem is timing.  

If it does not fix the issue, can you provide me with the exact model of the jewel neopixel you bought from Adafruit.  I'd like to get one as well so I can try to reproduce the issue and debug.

**Comment by johnwiseheart** [comment]
Aha, yes setting the `is_rgbw` flag did it for me - somehow I missed that in the docs. Hopefully it will solve for @imsadegh too

**Comment by HomeSpan** [comment]
That's really good news.  Thanks for the update.

**Comment by HomeSpan** [comment]
@imsadegh, did this solve your issue?

**Comment by imsadegh** [comment]
> but it will let us see if the problem is timing

Changing the timing with the below code didn't work.
`pixel->setTiming(0.40, 0.85, 0.80, 0.45, 80);`

And this is my 4x4 RGB led [buy link](https://www.aliexpress.us/item/2251832458182101.html?gatewayAdapt=tur2usa4itemAdapt).

**Comment by imsadegh** [comment]
In another LED strip with 175 LEDs, it didn't flicks on all leds. But the colors are not correct after some random LEDs
[This is](https://ibb.co/bXNcYXr) the Strip LEDs after I set it to red with Siri. And these are [Green](https://ibb.co/pdzCFn4) and [Blue](https://ibb.co/yhDnGq0).

**Comment by HomeSpan** [comment]
Unfortunately I'm not sure how to debug this issue. What ESP are you using?

**Comment by HomeSpan** [comment]
And to confirm, you checked you are using the correct flag for RGB versus RGBW?

**Comment by imsadegh** [comment]
> What ESP are you using

It's ESP32 version 1, I also tried it with the ESP-C3 with the same result.
I just upload the code and connect the my ws2812B to the pin where in the code named RGB pin.

**Comment by imsadegh** [comment]
I wanted to try it with FastLED, But it was too complicated for me to implement.😅

**Comment by imsadegh** [comment]
I believe that the problem is caused by the timing.

**Comment by imsadegh** [comment]
I modified the code to work with the FastLed lib, This is the code:

```
#if defined(CONFIG_IDF_TARGET_ESP32)
#define NEOPIXEL_RGB_PIN 5
#define NEOPIXEL2_RGB_PIN 17
#define DEVICE_SUFFIX ""

#endif


#include "HomeSpan.h"
#include <FastLED.h>


struct NeoPixel_RGB : Service::LightBulb {
  Characteristic::On power{ 0, true };
  Characteristic::Hue H{ 0, true };
  Characteristic::Saturation S{ 0, true };
  Characteristic::Brightness V{ 100, true };
  CRGB *leds;
  uint8_t nPixels;

  NeoPixel_RGB(uint8_t pin, int nPixels)
    : Service::LightBulb() {
    V.setRange(5, 100, 1);
    
    FastLED.addLeds<WS2812B, NEOPIXEL_RGB_PIN, GRB>(leds, nPixels);
    this->nPixels = nPixels;

    update();
  }

  boolean update() override {
    int p = power.getNewVal();
    float h = H.getNewVal<float>();
    float s = S.getNewVal<float>();
    float v = V.getNewVal<float>();

    CRGB color = CHSV(h * p, s * p, v * p);
    fill_solid(leds, nPixels, color);
    FastLED.show();

    return (true);
  }
};

void setup() {
  Serial.begin(115200);

  homeSpan.setPairingCode("11122333");
  homeSpan.setQRID("111-22-333");
  homeSpan.setWifiCredentials("HDP", "11336699");
  homeSpan.begin(Category::Lighting, "Pixel RGB Hub" DEVICE_SUFFIX);

  SPAN_ACCESSORY();

  SPAN_ACCESSORY("RGB");
  new NeoPixel_RGB(NEOPIXEL_RGB_PIN, 350);
}

void loop() {
  homeSpan.poll();
}
```


After running it, it will keep resetting the ESP.
